File: Python/Trabalho_POO_parte_2/Carro_Esportivo.py
from Veiculo_Motorizado import *
from Veiculo import *
from Interface_IPVA import *


class Carro_Esportivo(Veiculo_Motorizado, Interface_IPVA):
    __gasto_carro_esportivo: float = 2.3
    # constructor
    def __init__(self, id: int, rodas: int):
        super().__init__(id, rodas)
        super().set_valor_ipva(self.calcular_ipva())
        self.__tipo = "E"
        print(f"Carro Esportivo com ID {self.get_id()} foi construida")

    # metodo get
    def get_tipo(self) -> int:
        return self.__tipo

    # funções abstrata herdada pelo pai veiculo
    def mover(self):
        if self.get_ipva():
            if self.get_combustivel() >= self.__gasto_carro_esportivo:
                if self.verificar_pneu():
                    self.consumir(self.__gasto_carro_esportivo)
                    new_distancia = self.get_distancia_percorrida() + 2
                    self.set_distancia_percorrida(new_distancia)
                    print(f"Carro Esportivo - ID:{self.get_id()} andou  bloco +10 e está com {self.get_combustivel():.2f}litros de gasolina")
                else:
                    print(f"Carro Esportivo com ID {self.get_id()} não possui todos os pneus calibrados")
            else:
                print(f"Carro Esportivo com ID {self.get_id()} não possui combustivel suficiente")
        else:
            print(f"Carro Esportivo com ID {self.get_id()} não está com ipva em dia")

    def desenhar(self):
        print(" "*self.get_distancia_percorrida() + "        __")
        print(" "*self.get_distancia_percorrida() + "      ~( @\ \ ")
        print(" "*self.get_distancia_percorrida() + "   _____]_[_/_>__")
        print(" "*self.get_distancia_percorrida() + "  / __ \\<> |  __ \\")
        print(" "*self.get_distancia_percorrida() + "=\\_/__\\_\\__|_/__\\")
        print(" "*self.get_distancia_percorrida() + "   (__)      (__)")

    def calcular_ipva(self):
        return self.VALOR_BASE * self.CARRO_ESPORTIVO

File: C++/Trabalho_POO_parte_2/Carro_Esportivo.cpp
#include "Carro_Esportivo.h"

Carro_Esportivo::Carro_Esportivo(int id, int rodas) : Veiculo_Motorizado(id, rodas), tipo('E')
{
    tipo = 'E';
    std::cout << "Carro Esportivo com ID " << get_id() << " foi construído" << endl;
}

char Carro_Esportivo::get_tipo()
{
    return tipo;
}

void Carro_Esportivo::mover()
{
    if (get_ipva())
    {
        if (get_combustivel() >= gasto_carro_esportivo)
        {
            if (verificar_pneu())
            {
                consumir(gasto_carro_esportivo);
                int new_distancia = get_distancia_percorrida() + 10;
                set_distancia_percorrida(new_distancia);
                std::cout << "Carro Esportivo - ID:" << get_id() << " andou  bloco +10 e está com " << get_combustivel() << " litros de gasolina" << endl;
            }
            else
                std::cout << "Carro Esportivo com ID " << get_id() << " não possui todos os pneus calibrados" << endl;
        }
        else
            std::cout << "Carro Esportivo com ID " << get_id() << " não possui combustível suficiente" << endl;
    }
    else
        std::cout << "Carro Esportivo com ID " << get_id() << " não está com IPVA em dia" << endl;
}

void Carro_Esportivo::desenhar()
{
    string espacos(get_distancia_percorrida(), ' ');
    std::cout << espacos << "        __" << endl;
    std::cout << espacos << "      ~( @\\ \\" << endl;
    std::cout << espacos << "   _____]_[_/_>__" << endl;
    std::cout << espacos << "  / __ \\<> |  __ \\" << endl;
    std::cout << espacos << "=\\_/__\\_\\__|_/__\\" << endl;
    std::cout << espacos << "   (__)      (__)" << endl;
}
